### DISCUSSION TITLE
Bulk Import Metadata How?

### DISCUSSION THREAD

**Discussion by Dan1ell** [mediacms-io/mediacms#92] [Q&A]
Let’s say I invite a Youtuber to add some of their videos to my MediaCMS, let’s say 25 videos.

The Youtuber can easily upload the 25 videos from his/her hard drive, but what’s the best way to get their metadata in? So the title, description, category, tags, and perhaps the thumbnail image of each video. Adding that by hand for 25 videos is a lot to ask.

Ideas?

**Comment by mgogoulos** [comment]
Just a few thoughts on this

a) ability to import from Youtube, using the excellent youtube-dl utility. Not sure if this brings metadata too, but the Youtube Data API could be used in this case (this could add some complexity, because it requires adding an API key most probably, but is worth the effort)

b) add the media content using the REST API, which is not documented (it's a shame, I know). 
This is the way to go apparently, either through using a generic utility, or ideally using a MediaCMS client

Example using httpie client on demo.mediacms.io

```
user@user:~/Desktop$ http -a user:password -f POST https://demo.mediacms.io/api/v1/media/ media_file@video_x.mp4 title='the title' description='the description'
```
The REST API endpoint is /api/v1/media/ and there is a limitation now, it won't accept all metadata (so no category and tags), but it is something that can be added in the future. 

So the following actions need take place: 

a) document the API, so it can be used easily
b) build a client, that communicates to the API
c) add support for tags/category on the add media API call

d) integrate youtube-dl and allow addition of Youtube media by URL

**Reply by Dan1ell** [reply]
youtube-dl looks nice.
I'm thinking:

**Frontend**:
input [video or playlist url]
select [category]

**Backend**:
If video url, fetch video name.
If playlist url, fetch list of video names.

For each video name: check if video already present,
If present skip, else:

1. fetch video file, description, tags, and subtitles.
2. store video file, store subtitle files.
3. add category, description, tags.

**Reply by swiftugandan** [reply]
> Just a few thoughts on this
> 
> a) ability to import from Youtube, using the excellent youtube-dl utility. Not sure if this brings metadata too, but the Youtube Data API could be used in this case (this could add some complexity, because it requires adding an API key most probably, but is worth the effort)
> 
> b) add the media content using the REST API, which is not documented (it's a shame, I know).
> This is the way to go apparently, either through using a generic utility, or ideally using a MediaCMS client
> 
> Example using httpie client on demo.mediacms.io
> 
> ```
> user@user:~/Desktop$ http -a user:password -f POST https://demo.mediacms.io/api/v1/media/ media_file@video_x.mp4 title='the title' description='the description'
> ```
> 
> The REST API endpoint is /api/v1/media/ and there is a limitation now, it won't accept all metadata (so no category and tags), but it is something that can be added in the future.
> 
> So the following actions need take place:
> 
> a) document the API, so it can be used easily
> b) build a client, that communicates to the API
> c) add support for tags/category on the add media API call
> 
> d) integrate youtube-dl and allow addition of Youtube media by URL

The api could be documented using openapi (swagger) using something like https://github.com/tfranzel/drf-spectacular

**Comment by gam-ragnar** [comment]
> Let’s say I invite a Youtuber to add some of their videos to my MediaCMS, let’s say 25 videos.
> 
> The Youtuber can easily upload the 25 videos from his/her hard drive, but what’s the best way to get their metadata in? So the title, description, category, tags, and perhaps the thumbnail image of each video. Adding that by hand for 25 videos is a lot to ask.
> 
> Ideas?

@Dan1ell I like this feature, I will donate to a bounty to have this implemented

**Reply by Dan1ell** [reply]
I'm happy to add a bounty as well.

**Comment by DreamShaded** [comment]
Hello! I Found a [reddit discussion](https://www.reddit.com/r/selfhosted/comments/pagkh8/comment/ha8jn2b/?utm_source=reddit&utm_medium=web2x&context=3) about mass import local files to MediaCMS, but coudn't find the way to do it. Is API the only one method to bulk import, so no metadata (tags, other) will be loaded?

**Comment by Dabitch** [comment]
So just like @Dan1ell I would be happy to join the pool making a bounty for this feature if possible.

**Comment by cmuc24** [comment]
as i have similar requirements, and can confirm, 
the official way for 'bulk uploads' covers only a minimum of common requirements.

If someone has no idea ...  **Official Bulk Uploads** ...can be done via [CLI-Tool](https://github.com/mediacms-io/mediacms/blob/main/cli-tool/README.md) 
Just follow the documentation (see link), and it should work as expected. 
**Tip**: Create a local 'upload_media' folder and try `python cli.py upload-media` 

Actual, i get no further but my customized API looks good so far, for a better bulk upload experience :)
- original videoIDs used for friendly_token
- original title, description, upload_date, views, likes, dislikes, comments provided by json files
- get all videoIDs from a YoutubeChannel and choose, which ones should be imported, or all

Missing Features
- Assign Mediafiles to another user
at the moment: creating users manual, separating videos/uploads, and each time logout/login
- Assign Channelbanner to users Channel
- API Support for creating new users/channels
- API Support for assigning categories and tags to mediafiles

For enhancing API, i'm in 👍
